sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsReadableSpan.java

```diff
@@ -516,13 +515,15 @@ int getTotalRecordedLinks() {
   }
 
   @GuardedBy("lock")
-  private Map<String, AttributeValue> getImmutableAttributes() {
+  private Attributes getImmutableAttributes() {
```

Comment:
Looking at this I see a lot of allocations happening here. Do you have a plan to change AttributeMap to implement the Attributes so we don't reconstruct the entire map when converting to SpanData?

Reply:
I would definitely like to tackle this next. So far, I'm a bit stumped on how to approach it, since the AttributeMap is extremely mutable, and Attributes are inherently immutable. Do you have any ideas of how to tackle this one?

Reply:
AttributeMap becomes immutable after end event so we can reuse that object

Reply:
ok, I've been noodling on this a bit.
If we made an exportable-attributes interface (no idea what the name of this could be), then both our existing Attributes abstract class and the AttributesMap class could both implement that interface. Then, SpanData would expose this interface, rather than the concrete Attributes API class. Is this the sort of thing you're also thinking?

Reply:
Naming wise, here is my suggestion: 

* Rename the current `Attributes` class to `ImmutableAttributes`, keeping the implementation and the static creation methods there.
* Create a new interface called `Attributes` which will be used as the API interface and for export as well.
* Make the AttributesMap implement the Attributes interface
* Do as you suggest and pass in the AttributeMap directly to the SpanWrapper after the span has ended.

Does this sound right?

Reply:
Note, I tried this out in a branch, and for the API it ends up feeling parallel to the Map/ImmutableMap APIs :  Attributes/ImmutableAttributes.

Reply:
A small question: will `ImmutableAttributes` implement `Attributes` as well?

Reply:
yes, ImmutableAttributes and AttributesMap will both implement the Attributes interface.